File: src/main/java/de/hasenburg/geofencebroker/model/payload/AbstractPayload.java
package de.hasenburg.geofencebroker.model.payload;

import com.fasterxml.jackson.annotation.JsonIgnore;
import de.hasenburg.geofencebroker.model.InternalClientMessage;
import de.hasenburg.geofencebroker.model.InternalBrokerMessage;
import de.hasenburg.geofencebroker.model.JSONable;

import java.lang.reflect.Field;
import java.util.Optional;

/**
 * Every {@link InternalClientMessage} or {@link InternalBrokerMessage} has at least this empty payload.
 */
public abstract class AbstractPayload implements JSONable {

	@SuppressWarnings("BooleanMethodIsAlwaysInverted")
	public boolean nullField() {
		try {
			for (Field f : getClass().getDeclaredFields()) {
				if (f.get(this) == null) {
					return true;
				}
			}
			return false;
		} catch (IllegalAccessException e) {
			logger.error("Exception while null checking fields", e);
			return true;
		}
	}

	/*****************************************************************
	 * Subclasses
	 ****************************************************************/

	@JsonIgnore
	public Optional<CONNACKPayload> getCONNACKPayload() {
		if (this instanceof CONNACKPayload) {
			return Optional.of((CONNACKPayload) this);
		} else {
			return Optional.empty();
		}
	}

	@JsonIgnore
	public Optional<CONNECTPayload> getCONNECTPayload() {
		if (this instanceof CONNECTPayload) {
			return Optional.of((CONNECTPayload) this);
		} else {
			return Optional.empty();
		}
	}

	@JsonIgnore
	public Optional<DISCONNECTPayload> getDISCONNECTPayload() {
		if (this instanceof DISCONNECTPayload) {
			return Optional.of((DISCONNECTPayload) this);
		} else {
			return Optional.empty();
		}
	}

	@JsonIgnore
	public Optional<PINGREQPayload> getPINGREQPayload() {
		if (this instanceof PINGREQPayload) {
			return Optional.of((PINGREQPayload) this);
		} else {
			return Optional.empty();
		}
	}

	@JsonIgnore
	public Optional<PINGRESPPayload> getPINGRESPPayload() {
		if (this instanceof PINGRESPPayload) {
			return Optional.of((PINGRESPPayload) this);
		} else {
			return Optional.empty();
		}
	}

	@JsonIgnore
	public Optional<PUBACKPayload> getPUBACKPayload() {
		if (this instanceof PUBACKPayload) {
			return Optional.of((PUBACKPayload) this);
		} else {
			return Optional.empty();
		}
	}

	@JsonIgnore
	public Optional<PUBLISHPayload> getPUBLISHPayload() {
		if (this instanceof PUBLISHPayload) {
			return Optional.of((PUBLISHPayload) this);
		} else {
			return Optional.empty();
		}
	}

	@JsonIgnore
	public Optional<SUBACKPayload> getSUBACKPayload() {
		if (this instanceof SUBACKPayload) {
			return Optional.of((SUBACKPayload) this);
		} else {
			return Optional.empty();
		}
	}

	@JsonIgnore
	public Optional<SUBSCRIBEPayload> getSUBSCRIBEPayload() {
		if (this instanceof SUBSCRIBEPayload) {
			return Optional.of((SUBSCRIBEPayload) this);
		} else {
			return Optional.empty();
		}
	}

	/*****************************************************************
	 * JSON and String
	 ****************************************************************/

	@Override
	public String toString() {
		return JSONable.toJSON(this);
	}

}

File: GeoBroker-Commons/src/main/kotlin/de/hasenburg/geobroker/commons/model/message/payloads/AbstractPayload.kt
package de.hasenburg.geobroker.commons.model.message.payloads

/**
 * Every [InternalServerMessage], [InternalBrokerMessage] and [InternalClientMessage] has at least this empty payload.
 */
open class AbstractPayload {

    /*****************************************************************
     * Subclasses
     ****************************************************************/

    fun getCONNECTPayload(): CONNECTPayload? {
        if (this is CONNECTPayload) return this else return null
    }

    fun getCONNACKPayload(): CONNACKPayload? {
        if (this is CONNACKPayload) return this else return null
    }

    fun getDISCONNECTPayload(): DISCONNECTPayload? {
        if (this is DISCONNECTPayload) return this else return null
    }

    fun getPINGREQPayload(): PINGREQPayload? {
        if (this is PINGREQPayload) return this else return null
    }

    fun getPINGRESPPayload(): PINGRESPPayload? {
        if (this is PINGRESPPayload) return this else return null
    }

    fun getPUBLISHPayload(): PUBLISHPayload? {
        if (this is PUBLISHPayload) return this else return null
    }

    fun getPUBACKPayload(): PUBACKPayload? {
        if (this is PUBACKPayload) return this else return null
    }

    fun getSUBSCRIBEPayload(): SUBSCRIBEPayload? {
        if (this is SUBSCRIBEPayload) return this else return null
    }

    fun getSUBACKPayload(): SUBACKPayload? {
        if (this is SUBACKPayload) return this else return null
    }

    fun getUNSUBSCRIBEPayload(): UNSUBSCRIBEPayload? {
        if (this is UNSUBSCRIBEPayload) return this else return null
    }

    fun getUNSUBACKPayload(): UNSUBACKPayload? {
        if (this is UNSUBACKPayload) return this else return null
    }

    fun getBrokerForwardDisconnectPayload(): BrokerForwardDisconnectPayload? {
        if (this is BrokerForwardDisconnectPayload) return this else return null
    }

    fun getBrokerForwardPingreqPayload(): BrokerForwardPingreqPayload? {
        if (this is BrokerForwardPingreqPayload) return this else return null
    }

    fun getBrokerForwardSubscribePayload(): BrokerForwardSubscribePayload? {
        if (this is BrokerForwardSubscribePayload) return this else return null
    }

    fun getBrokerForwardUnsubscribePayload(): BrokerForwardUnsubscribePayload? {
        if (this is BrokerForwardUnsubscribePayload) return this else return null
    }

    fun getBrokerForwardPublishPayload(): BrokerForwardPublishPayload? {
        if (this is BrokerForwardPublishPayload) return this else return null
    }

}
